File: DMOJ&WLMOJ Java/Frog1/src/frog1/Frog1.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package frog1;

import java.util.*;
import java.io.*;

/**
 *
 * @author Joshua
 */


public class Frog1 {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    /**
     * @param args the command line arguments
     */
    static int a[];
    static int dp[];
    static int size;
    public static void main(String[] args) throws Exception{
        // TODO code application logic here
        size = readInt();
        dp = new int[size];
        a = new int[size];
        for(int i=0; i<size; i++){
            dp[i]=-1;
            a[i]=readInt();
        }
        System.out.println(f(0));
    }
    static int f(int b){
        if(dp[b]!=-1)return dp[b];
        if(b==size-1)return 0;
        if(b+1==size-1)return Math.abs(a[b]-a[b+1])+f(b+1);
        if(Math.abs(a[b]-a[b+1])+f(b+1)<Math.abs(a[b]-a[b+2])+f(b+2)){
            dp[b]=Math.abs(a[b]-a[b+1])+f(b+1);
            return dp[b];
        }else{
            dp[b]=Math.abs(a[b]-a[b+2])+f(b+2);
            return dp[b];
        }
    }

    static String next () throws IOException {
        if (st == null || !st.hasMoreTokens())
            st = new StringTokenizer(br.readLine());
        return st.nextToken();
    }

    static long readLong() throws IOException {
        return Long.parseLong(next());
    }

    static int readInt() throws IOException {
         return Integer.parseInt(next());
    }

    static double readDouble() throws IOException {
         return Double.parseDouble(next());
    }

    static char readCharacter() throws IOException {
          return next().charAt(0);
    }

    static String readLine() throws IOException {
        return br.readLine().trim();
    }
}

File: DMOJ&MCPT/Frog1/Frog1.cpp
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/* 
 * File:   main.cpp
 * Author: Joshua
 *
 * Created on April 10, 2019, 6:20 p.m.
 */

#include <string>
#include <bits/stdc++.h>

using namespace std;

/*
 * 
 */
int a[100001];
int dp[100001];
int size;
int k;
int f(int b){
    if(dp[b]!=-1)return dp[b];
    if(b==size-1)return 0;
    int min=INT_MAX;
    int d = k<size-b-1?k:size-b-1;
    for(int i=1; i<=d; i++){
        int c = abs(a[b]-a[b+i])+f(b+i);
        min=c<min?c:min;
    }
    dp[b]=min;
    return dp[b];
}
int main() {
    
    scanf("%d %d",&size,&k);
    memset(dp,-1,sizeof(dp));
    for(int i=0; i<size; i++){
        scanf("%d\n",&a[i]);
        dp[i]=-1;
    }
    printf("%d",f(0));

}

